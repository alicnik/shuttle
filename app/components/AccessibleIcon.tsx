import * as AccessibleIconComponent from '@radix-ui/react-accessible-icon';

export const AccessibleIcon = AccessibleIconComponent.Root;
