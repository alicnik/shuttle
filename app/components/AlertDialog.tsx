import {
  AlertDialog as AlertDialogComponent,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
} from '~/components/ui/alert-dialog';

type AlertDialogProps = React.PropsWithChildren<{
  title: string;
  description: string;
  onConfirm: () => void;
}>;

export function AlertDialog({
  children,
  title,
  description,
  onConfirm,
}: AlertDialogProps) {
  return (
    <AlertDialogComponent>
      {children}
      <AlertDialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <AlertDialogHeader>
          <AlertDialogTitle>{title}</AlertDialogTitle>
          <AlertDialogDescription>{description}</AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>Cancel</AlertDialogCancel>
          <AlertDialogAction onClick={onConfirm}>Continue</AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialogComponent>
  );
}
