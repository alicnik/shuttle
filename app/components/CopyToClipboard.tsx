import { ClipboardCopyIcon } from '@radix-ui/react-icons';
import { useToast } from './ui';
import { Tooltip } from './Tooltip';
import type { TooltipProps } from './Tooltip';

interface CopyToClipboardProps {
  copyText: string;
  successMessage: string;
  tooltipOptions: Omit<TooltipProps, 'children'>;
}

export function CopyToClipboard({
  copyText,
  successMessage,
  tooltipOptions: { content, side = 'top', sideOffset = 0 },
}: CopyToClipboardProps) {
  const { toast } = useToast();

  return (
    <Tooltip content={content} side={side} sideOffset={sideOffset}>
      <ClipboardCopyIcon
        className="translate-y-[3px] cursor-pointer"
        onClick={async () => {
          await navigator.clipboard.writeText(copyText);
          toast({ description: successMessage });
        }}
      />
    </Tooltip>
  );
}
