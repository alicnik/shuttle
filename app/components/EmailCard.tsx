import type { Email } from '@prisma/client';
import { Checkbox } from '~/components/ui/checkbox';
import { useNavigation, useSubmit } from '@remix-run/react';
import { formatRelative } from 'date-fns';

interface EmailCardProps {
  email: Email;
  selected: string[];
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
  preview: string | null;
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
}

export function EmailCard({
  email,
  selected,
  setSelected,
  preview,
  setPreview,
}: EmailCardProps) {
  const submit = useSubmit();
  const navigation = useNavigation();

  if (navigation.state === 'submitting') {
    console.log(Object.fromEntries(navigation.formData || []));
  }

  const isOptimisticRead =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markRead' &&
    navigation.formData?.get('emailId') === email.id;

  const isOptimisticUnread =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markSelectedUnread' &&
    String(navigation.formData?.get('selected')).split(',').includes(email.id);

  const isSelected = selected.includes(email.id);
  const isPreview = preview === email.id;

  const from = email.from.startsWith('<')
    ? email.from.slice(1, -1)
    : email.from.split('<')[0].trim();

  return (
    <div
      data-selected={isSelected}
      data-in-view={isPreview}
      className="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-800 data-[in-view=true]:bg-zinc-800 data-[selected=true]:bg-zinc-800"
    >
      <Checkbox
        className="mt-[2px]"
        checked={isSelected || isPreview}
        onCheckedChange={(checked) => {
          setSelected((selected) => {
            if (checked) {
              return [...selected, email.id];
            }

            if (isPreview) {
              setPreview(null);
            }
            return selected.filter((id) => id !== email.id);
          });
        }}
      />
      <div
        data-read={!isOptimisticUnread && (email.read || isOptimisticRead)}
        className="cursor-pointer text-left data-[read=true]:opacity-50"
        onClick={() =>
          setPreview((current) => {
            if (current) {
              submit(
                { emailId: current, _action: 'markRead' },
                { method: 'post' }
              );
            }
            return email.id;
          })
        }
      >
        <div className="mb-1 flex items-center ">
          <p className="mr-12 w-max text-sm font-bold">{from}</p>
          <p className="ml-auto w-max text-xs">
            {formatRelative(email.createdAt, new Date())}
          </p>
        </div>
        <p className="mb-1 text-sm">{email.subject}</p>
        <p className="line-clamp-2 text-xs">{email.text}</p>
      </div>
    </div>
  );
}
