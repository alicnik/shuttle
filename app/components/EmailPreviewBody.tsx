import clsx from 'clsx';
import type { Email } from '@prisma/client';
import * as React from 'react';

export function EmailPreviewBody({ html }: { html?: Email['html'] }) {
  const htmlWithBaseTag = React.useMemo(() => {
    if (!html) return '';

    // This ensures links open in a new tab and not in the iframe
    const baseTag = '<base target="_blank">';
    const headTag = '<head>';
    const headTagIndex = html.indexOf(headTag);

    if (headTagIndex === -1) {
      return html;
    }

    return (
      html.slice(0, headTagIndex + headTag.length) +
      baseTag +
      html.slice(headTagIndex + headTag.length)
    );
  }, [html]);

  return (
    <iframe
      data-testid="email-preview-body"
      className={clsx('email-preview', 'h-full w-full')}
      srcDoc={htmlWithBaseTag}
      title="Email preview"
    />
  );
}
