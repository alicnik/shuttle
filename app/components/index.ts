export * from './AlertDialog';
export * from './CopyToClipboard';
export * from './EmailCard';
export * from './EmailPreviewBody';
export * from './EmailPreviewHeader';
export * from './HarvestLinks';
export * from './Theme';
export * from './Tooltip';
export * from './ui';
