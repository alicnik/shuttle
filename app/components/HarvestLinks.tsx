import * as React from 'react';
import {
  Button,
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
  useToast,
} from './ui';
import { Tooltip } from './Tooltip';
import { Link2Icon } from '@radix-ui/react-icons';
import { LINK_COPY_SUCCESS_MESSAGE } from '~/lib';
import type { Email } from '@prisma/client';
import { AccessibleIcon } from './AccessibleIcon';

interface HarvestLinksProps {
  email: Email;
}

export function HarvestLinks({ email }: HarvestLinksProps) {
  const { toast } = useToast();

  const links = React.useMemo(() => {
    /* c8 ignore next 3 */
    if (typeof window === 'undefined') {
      return [];
    }
    const parser = new DOMParser();
    const doc = parser.parseFromString(email.html, 'text/html');
    const links = Array.from(doc.querySelectorAll('a'));
    const mapped = links.map((link) => {
      const linkDoc = parser.parseFromString(link.innerHTML, 'text/html');
      const elements = linkDoc.querySelectorAll('*');
      elements.forEach((element) => {
        element.removeAttribute('style');
        element.removeAttribute('width');
        element.removeAttribute('height');
        if (element.tagName === 'IMG') {
          element.setAttribute('style', 'max-width: 80%;');
        }
      });
      link.innerHTML = linkDoc.body.innerHTML;
      return link;
    });
    return mapped;
  }, [email.html]);

  return (
    <Dialog>
      <Tooltip content="Harvest links">
        <DialogTrigger asChild>
          <Button
            variant="ghost"
            className="hover:bg-zinc-300"
            type="button"
            size="icon"
          >
            <AccessibleIcon label="Harvest links from email">
              <Link2Icon />
            </AccessibleIcon>
          </Button>
        </DialogTrigger>
      </Tooltip>
      <DialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <DialogHeader>
          <DialogTitle>Found the following links in the email</DialogTitle>
          <DialogDescription>
            Click a link to copy it to your clipboard
          </DialogDescription>
          {links.length > 0 ? (
            <ul className="grid auto-rows-fr grid-cols-2 place-items-center gap-x-2 gap-y-8 pt-8">
              {links.map((link) => {
                return (
                  <Tooltip key={link.href} content={link.href} sideOffset={4}>
                    <li
                      className="flex cursor-pointer justify-center"
                      dangerouslySetInnerHTML={{
                        __html: link.innerHTML,
                      }}
                      onClick={async () => {
                        await navigator.clipboard.writeText(link.href);
                        toast({
                          description: LINK_COPY_SUCCESS_MESSAGE,
                        });
                      }}
                    />
                  </Tooltip>
                );
              })}
            </ul>
          ) : (
            <DialogDescription>
              No links were found in the email
            </DialogDescription>
          )}
        </DialogHeader>
      </DialogContent>
    </Dialog>
  );
}
