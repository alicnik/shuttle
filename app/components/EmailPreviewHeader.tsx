import { Form, Link, useSubmit } from '@remix-run/react';
import { format } from 'date-fns';
import { Tooltip } from './Tooltip';
import { Button, Separator } from './ui';
import type { Email, User } from '@prisma/client';
import {
  EnvelopeClosedIcon,
  ExternalLinkIcon,
  TrashIcon,
} from '@radix-ui/react-icons';
import { AlertDialog } from './AlertDialog';
import { AlertDialogTrigger } from './ui/alert-dialog';

interface EmailPreviewHeaderProps {
  email: Email;
  userId: User['id'];
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
}

export function EmailPreviewHeader({
  email,
  userId,
  setPreview,
  setSelected,
}: EmailPreviewHeaderProps) {
  const submit = useSubmit();

  return (
    <div className="max-h-48 overflow-hidden p-4 text-zinc-900">
      <h3 className="truncate">
        <span className="font-bold">From:</span> {email.from}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">To:</span> {`${email.userId}@shuttle.email`}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">Subject:</span> {email.subject}
      </h3>
      <h3 className="mb-4 truncate">
        <span className="font-bold">Received:</span>{' '}
        {format(email.createdAt, 'cccc, io MMMM yyyy @ hh:mm a')}
      </h3>
      <Form method="post" className="mb-4 flex gap-2">
        <input type="hidden" name="selected" value={email.id} />
        <Tooltip content="Mark as unread">
          <Button
            variant="ghost"
            size="icon"
            name="_action"
            value="markSelectedUnread"
          >
            <EnvelopeClosedIcon />
          </Button>
        </Tooltip>
        <AlertDialog
          title="Are you sure?"
          description="This action cannot be undone."
          onConfirm={() => {
            setPreview(null);
            setSelected((selected) => selected.filter((id) => id !== email.id));
            submit(
              { selected: email.id, _action: 'deleteSelected' },
              { method: 'post' }
            );
          }}
        >
          <Tooltip content="Delete">
            <AlertDialogTrigger asChild>
              <Button
                name="_action"
                value="deleteSelected"
                variant="ghost"
                size="icon"
                className="hover:bg-destructive focus:bg-destructive"
              >
                <TrashIcon />
              </Button>
            </AlertDialogTrigger>
          </Tooltip>
        </AlertDialog>
        <Tooltip content="Open in new window">
          <Link to={`/${userId}/${email.id}`} target="_blank">
            <Button
              type="button"
              variant="ghost"
              size="icon"
              onClick={() => {
                submit(
                  { emailId: email.id, _action: 'markRead' },
                  { method: 'post' }
                );
              }}
            >
              <ExternalLinkIcon />
            </Button>
          </Link>
        </Tooltip>
      </Form>
      <Separator />
    </div>
  );
}
