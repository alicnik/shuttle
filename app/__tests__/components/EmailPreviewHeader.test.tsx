import { screen } from '@testing-library/react';
import { EmailPreviewHeader } from '~/components';
import { setupRemixStub } from '../test-utils';
import { useSearchParams, useSubmit } from '@remix-run/react';

const props = {
  email: {
    id: 'mock-id',
    subject: 'mock-subject',
    from: 'mock-from',
    text: 'mock-text',
    html: 'mock-html',
    userId: 'mock-user-id',
    read: false,
    createdAt: new Date('October 13, 2023 13:07:00'),
  },
  setSelected: vi.fn(),
};

describe('EmailPreviewHeader', () => {
  it('should match the snapshot', () => {
    const { container } = setupRemixStub(<EmailPreviewHeader {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should render the email details', () => {
    setupRemixStub(<EmailPreviewHeader {...props} />);

    const from = screen.getByRole('heading', { name: /from: mock-from/i });
    const to = screen.getByRole('heading', {
      name: /to: mock-user-id@shuttle.email/i,
    });
    const subject = screen.getByRole('heading', {
      name: /subject: mock-subject/i,
    });
    const received = screen.getByRole('heading', {
      name: /received: friday, 13th october 2023 @ 01:07 pm/i,
    });

    expect(from).toHaveTextContent('From: mock-from');
    expect(to).toHaveTextContent('To: mock-user-id@shuttle.email');
    expect(subject).toHaveTextContent('Subject: mock-subject');
    expect(received).toHaveTextContent(
      'Received: Friday, 13th October 2023 @ 01:07 PM'
    );
  });

  it('should update state upon deletion and programmatically submit form values', async () => {
    const mockSearchParams = new URLSearchParams({ preview: props.email.id });
    const mockSetSearchParams = vi.fn();
    vi.mocked(useSearchParams).mockReturnValueOnce([
      mockSearchParams,
      mockSetSearchParams,
    ]);

    const mockSubmit = vi.fn();
    vi.mocked(useSubmit).mockReturnValueOnce(mockSubmit);

    const { user } = setupRemixStub(<EmailPreviewHeader {...props} />);

    await user.click(screen.getByText(/delete/i));
    await user.click(screen.getByRole('button', { name: /continue/i }));

    const setSearchParamsCallback = mockSetSearchParams.mock.lastCall[0];
    const newSearchParams = setSearchParamsCallback(mockSearchParams);

    const setSelectedCallback = props.setSelected.mock.lastCall[0];
    const newSelected = setSelectedCallback(['existing-id', props.email.id]);

    expect(newSearchParams.get('preview')).toBeNull();
    expect(newSelected).toEqual(['existing-id']);
    expect(mockSubmit).toHaveBeenCalledWith(
      { selected: props.email.id, _action: 'deleteSelected' },
      { method: 'post' }
    );
  });

  it('should mark the email as read when clicking on the link to open the email in a new window', async () => {
    const mockSubmit = vi.fn();
    vi.mocked(useSubmit).mockReturnValueOnce(mockSubmit);

    const { user } = setupRemixStub(<EmailPreviewHeader {...props} />);

    await user.click(screen.getByText(/open in new window/i));

    expect(mockSubmit).toHaveBeenCalledWith(
      { emailId: props.email.id, _action: 'markRead' },
      { method: 'post' }
    );
  });
});
