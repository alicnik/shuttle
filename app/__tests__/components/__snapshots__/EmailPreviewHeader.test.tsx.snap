// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailPreviewHeader > should match the snapshot 1`] = `
<div>
  <div
    class="max-h-48 overflow-hidden p-4 text-zinc-900"
  >
    <h3
      class="truncate"
    >
      <span
        class="font-bold"
      >
        From:
      </span>
       
      mock-from
    </h3>
    <h3
      class="truncate"
    >
      <span
        class="font-bold"
      >
        To:
      </span>
       
      mock-user-id@shuttle.email
    </h3>
    <h3
      class="truncate"
    >
      <span
        class="font-bold"
      >
        Subject:
      </span>
       
      mock-subject
    </h3>
    <h3
      class="mb-4 truncate"
    >
      <span
        class="font-bold"
      >
        Received:
      </span>
       
      Friday, 13th October 2023 @ 01:07 PM
    </h3>
    <form
      action="/"
      class="mb-4 flex gap-2"
      method="post"
    >
      <input
        name="selected"
        type="hidden"
        value="mock-id"
      />
      <button
        class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:text-accent-foreground h-9 w-9 hover:bg-zinc-300"
        data-state="closed"
        name="_action"
        value="markSelectedUnread"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M1 2C0.447715 2 0 2.44772 0 3V12C0 12.5523 0.447715 13 1 13H14C14.5523 13 15 12.5523 15 12V3C15 2.44772 14.5523 2 14 2H1ZM1 3L14 3V3.92494C13.9174 3.92486 13.8338 3.94751 13.7589 3.99505L7.5 7.96703L1.24112 3.99505C1.16621 3.94751 1.0826 3.92486 1 3.92494V3ZM1 4.90797V12H14V4.90797L7.74112 8.87995C7.59394 8.97335 7.40606 8.97335 7.25888 8.87995L1 4.90797Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <span
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        >
          Mark as unread
        </span>
      </button>
      <button
        aria-controls="radix-:r1:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:text-accent-foreground h-9 w-9 hover:bg-destructive"
        data-state="closed"
        name="_action"
        type="button"
        value="deleteSelected"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.5 1C5.22386 1 5 1.22386 5 1.5C5 1.77614 5.22386 2 5.5 2H9.5C9.77614 2 10 1.77614 10 1.5C10 1.22386 9.77614 1 9.5 1H5.5ZM3 3.5C3 3.22386 3.22386 3 3.5 3H5H10H11.5C11.7761 3 12 3.22386 12 3.5C12 3.77614 11.7761 4 11.5 4H11V12C11 12.5523 10.5523 13 10 13H5C4.44772 13 4 12.5523 4 12V4L3.5 4C3.22386 4 3 3.77614 3 3.5ZM5 4H10V12H5V4Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <span
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        >
          Delete
        </span>
      </button>
      <a
        data-state="closed"
        href="/mock-user-id/mock-id"
        target="_blank"
      >
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:text-accent-foreground h-9 w-9 hover:bg-zinc-300"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="none"
            focusable="false"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3 2C2.44772 2 2 2.44772 2 3V12C2 12.5523 2.44772 13 3 13H12C12.5523 13 13 12.5523 13 12V8.5C13 8.22386 12.7761 8 12.5 8C12.2239 8 12 8.22386 12 8.5V12H3V3L6.5 3C6.77614 3 7 2.77614 7 2.5C7 2.22386 6.77614 2 6.5 2H3ZM12.8536 2.14645C12.9015 2.19439 12.9377 2.24964 12.9621 2.30861C12.9861 2.36669 12.9996 2.4303 13 2.497L13 2.5V2.50049V5.5C13 5.77614 12.7761 6 12.5 6C12.2239 6 12 5.77614 12 5.5V3.70711L6.85355 8.85355C6.65829 9.04882 6.34171 9.04882 6.14645 8.85355C5.95118 8.65829 5.95118 8.34171 6.14645 8.14645L11.2929 3H9.5C9.22386 3 9 2.77614 9 2.5C9 2.22386 9.22386 2 9.5 2H12.4999H12.5C12.5678 2 12.6324 2.01349 12.6914 2.03794C12.7504 2.06234 12.8056 2.09851 12.8536 2.14645Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            Open in new window
          </span>
        </button>
      </a>
      <button
        aria-controls="radix-:r6:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:text-accent-foreground h-9 w-9 hover:bg-zinc-300"
        data-state="closed"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M8.51194 3.00541C9.18829 2.54594 10.0435 2.53694 10.6788 2.95419C10.8231 3.04893 10.9771 3.1993 11.389 3.61119C11.8009 4.02307 11.9513 4.17714 12.046 4.32141C12.4633 4.95675 12.4543 5.81192 11.9948 6.48827C11.8899 6.64264 11.7276 6.80811 11.3006 7.23511L10.6819 7.85383C10.4867 8.04909 10.4867 8.36567 10.6819 8.56093C10.8772 8.7562 11.1938 8.7562 11.389 8.56093L12.0077 7.94221L12.0507 7.89929C12.4203 7.52976 12.6568 7.2933 12.822 7.0502C13.4972 6.05623 13.5321 4.76252 12.8819 3.77248C12.7233 3.53102 12.4922 3.30001 12.1408 2.94871L12.0961 2.90408L12.0515 2.85942C11.7002 2.508 11.4692 2.27689 11.2277 2.11832C10.2377 1.46813 8.94398 1.50299 7.95001 2.17822C7.70691 2.34336 7.47044 2.57991 7.1009 2.94955L7.058 2.99247L6.43928 3.61119C6.24401 3.80645 6.24401 4.12303 6.43928 4.31829C6.63454 4.51355 6.95112 4.51355 7.14638 4.31829L7.7651 3.69957C8.1921 3.27257 8.35757 3.11027 8.51194 3.00541ZM4.31796 7.14672C4.51322 6.95146 4.51322 6.63487 4.31796 6.43961C4.12269 6.24435 3.80611 6.24435 3.61085 6.43961L2.99213 7.05833L2.94922 7.10124C2.57957 7.47077 2.34303 7.70724 2.17788 7.95035C1.50265 8.94432 1.4678 10.238 2.11799 11.2281C2.27656 11.4695 2.50766 11.7005 2.8591 12.0518L2.90374 12.0965L2.94837 12.1411C3.29967 12.4925 3.53068 12.7237 3.77214 12.8822C4.76219 13.5324 6.05589 13.4976 7.04986 12.8223C7.29296 12.6572 7.52943 12.4206 7.89896 12.051L7.89897 12.051L7.94188 12.0081L8.5606 11.3894C8.75586 11.1941 8.75586 10.8775 8.5606 10.6823C8.36533 10.487 8.04875 10.487 7.85349 10.6823L7.23477 11.301C6.80777 11.728 6.6423 11.8903 6.48794 11.9951C5.81158 12.4546 4.95642 12.4636 4.32107 12.0464C4.17681 11.9516 4.02274 11.8012 3.61085 11.3894C3.19896 10.9775 3.0486 10.8234 2.95385 10.6791C2.53661 10.0438 2.54561 9.18863 3.00507 8.51227C3.10993 8.35791 3.27224 8.19244 3.69924 7.76544L4.31796 7.14672ZM9.62172 6.08558C9.81698 5.89032 9.81698 5.57373 9.62172 5.37847C9.42646 5.18321 9.10988 5.18321 8.91461 5.37847L5.37908 8.91401C5.18382 9.10927 5.18382 9.42585 5.37908 9.62111C5.57434 9.81637 5.89092 9.81637 6.08619 9.62111L9.62172 6.08558Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
        <span
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        >
          Harvest links from email
        </span>
      </button>
    </form>
    <div
      class="shrink-0 bg-zinc-200 dark:bg-zinc-800 h-[1px] w-full"
      data-orientation="horizontal"
      role="none"
    />
  </div>
</div>
`;
